### PR TITLE
Options for PE and RHEL-H added

### PR DESCRIPTION
**Pull request by fftux**
Options were added to configure systems to use puppet enterprise (v3) instead of integrated puppet in katello.  Additional mods may need to be made to be v4 compliant.  

Option added to configure virt-who on RHEL based systems that are configured as RHV/oVirt hypervisors.

### CODE REVIEW

**Comment by evgeni** [comment]
Ohai,

I will add some inline comments, but here are a few general ones :-)
- can you sqash the changes in a few logical commits please?
- I don't think doing virt-who is bootstrap.py's job, we have Puppet/Ansible for that (but @sideangleside and @mccun934 might think otherwise)

File: bootstrap.py [hunk 1]
```diff
@@ -232,6 +236,11 @@ def register_systems(org_name, activationkey, release):
 def unregister_system():
     print_generic("Unregistering")
     exec_failexit("/usr/sbin/subscription-manager unregister")
+    
+    
+def enable_satellite_tools():
```
Comment (evgeni) [inline bootstrap.py:241]: that's a downstream Satellite thing, now replaced with `--enablerepos` upstream.

File: bootstrap.py [hunk 2]
```diff
@@ -288,11 +301,24 @@ def install_puppet_agent():
     exec_failexit("/sbin/service puppet restart")
 
 
+def install_pe_agent(org_name):
+    org_label = return_matching_katello_key('organizations', 'name="%s"' % org_name, 'label', False)
+    if os.path.isfile('/etc/yum.repos.d/pe_repo.repo'):
+        exec_failexit("mv /etc/yum.repos.d/pe_repo.repo{,.bak}")
+    exec_failexit("subscription-manager attach --pool=$(subscription-manager list --available --matches='Puppet Enterprise' --pool-only)" )
```
Comment (evgeni) [inline bootstrap.py:308]: the pool belongs into the activation-key, not into our calls

File: bootstrap.py [hunk 3]
```diff
@@ -288,11 +301,24 @@ def install_puppet_agent():
...
+    exec_failexit("subscription-manager repos --disable=*-pe-*")
+    exec_failexit("subscription-manager repos --enable=%s_Puppet_Enterprise_$(uname -r | cut -d. -f6)-pe-$(uname -r | cut -d. -f7)" % ( org_label ))
```
Comment (evgeni) [inline bootstrap.py:310]: those two should be done using `--enablerepos` as we don't know how the repos will be called at the users setup

File: bootstrap.py [hunk 4]
```diff
@@ -288,11 +301,24 @@ def install_puppet_agent():
...
+    print_generic("Installing the PE Agent")
+    yum("install", "pe-agent")
+    exec_failexit("/sbin/chkconfig pe-puppet on")
+    exec_failexit("/usr/local/bin/puppet agent -t --server %s || true" % (options.pe_server_fqdn))
```
Comment (evgeni) [inline bootstrap.py:314]: why `|| true`? this hides errors in puppet comms

File: bootstrap.py [hunk 5]
```diff
@@ -288,11 +301,24 @@ def install_puppet_agent():
...
+    exec_failexit("/sbin/service pe-puppet restart")
+
+
 def remove_obsolete_packages():
     print_generic("Removing old RHN packages")
     yum("remove", "rhn-setup rhn-client-tools yum-rhn-plugin rhnsd rhn-check rhnlib spacewalk-abrt spacewalk-oscap osad rh-*-rhui-client")
 
-
```
Comment (evgeni) [inline bootstrap.py:295]: pep8 wants this :)

File: bootstrap.py [hunk 6]
```diff
@@ -323,6 +349,20 @@ def install_foreman_ssh_key():
         os.chown(foreman_ssh_authfile, userpw.pw_uid, userpw.pw_gid)
         print_generic("Foreman's SSH key was added to %s" % foreman_ssh_authfile)
 
+def install_virt_who():
```
Comment (evgeni) [inline bootstrap.py:352]: as mentioned, I don't think this belongs here

File: bootstrap.py [hunk 7]
```diff
@@ -600,18 +640,29 @@ class MEOptions:
     register_systems(options.org, options.activationkey, options.release)
 
 if not options.remove:
+    enable_satellite_tools()
```
Comment (evgeni) [inline bootstrap.py:643]: Satellite stuff…

File: bootstrap.py [hunk 8]
```diff
@@ -288,11 +301,24 @@ def install_puppet_agent():
     exec_failexit("/sbin/service puppet restart")
 
 
+def install_pe_agent(org_name):
+    org_label = return_matching_katello_key('organizations', 'name="%s"' % org_name, 'label', False)
+    if os.path.isfile('/etc/yum.repos.d/pe_repo.repo'):
```
Comment (evgeni) [inline bootstrap.py:306]: I _think_ I'd leave that to the user. The file might be called differently too.

File: bootstrap.py [hunk 9]
```diff
@@ -288,11 +301,24 @@ def install_puppet_agent():
     exec_failexit("/sbin/service puppet restart")
 
 
+def install_pe_agent(org_name):
```
Comment (evgeni) [inline bootstrap.py:304]: can't we merge that and install_puppet? just having a flag whether it's puppet or PE that is going to be installed?

**Review by evgeni — CHANGES_REQUESTED** [review]

**Comment by evgeni** [comment]
Ah, and one more thing, we're missing a cleanup for PE installs.

File: bootstrap.py (hunk 2 above)
Comment (sideangleside) [reply bootstrap.py:308]: +1 Attaching subs is the activation keys' job. Also, this  assumes the custom Product is named **Puppet Enterprise** . 

Also, file operations (creates, deletes, moves & removes) should use native Python modules to remain consistent.

File: bootstrap.py (hunk 3 above)
Comment (sideangleside) [reply bootstrap.py:310]: Or done via the activation key. It is the user's choice, but -1 to hardcoded repo names.

**Comment by sideangleside** [comment]
I don't think a function to install (and configure) virt-who is within the scope of what bootstrap is designed to do. I would be open to, as a courtesy to users who are using Katello solely as a content source and have no config mgmt, a generic function to install additional packages (provided in a comma separated fashion). After that, we are well into Puppet/Ansible territory.

File: bootstrap.py (hunk 9 above)
Comment (sideangleside) [reply bootstrap.py:304]: Yes. Though the PE Agent requires a bit more configuration especially if a user is using Trusted Facts.

**Pull request closed without merging**